docs/design/rosbag2_playback_time.md

```diff
@@ -187,6 +187,53 @@ No special allowance needs to be made in the `Player` - time will be provided at
 
 The `Player` must register a `JumpHandler` with the `Clock` - so that when a jump occurs, the current message playback queue can be invalidated and re-enqueued according to the new starting time.
 
+## Time Synchronization and Conversion
+
+When implementing this time control, there are two separate time streams to consider:
+1. "ROS time": this is the time that is related to the messages in the bag - this time may speed up, slow down, pause, or jump based on rosbag playback controls (or external Time Sources)
+2. "Steady time": this time always moves forward, and always moves at the same rate; it is the real time that the user watching a playback experiences
+
+To implement `now() -> ROSTime` and `sleep_until(ROSTime until)`, we need to perform arbitrary conversion between "ROS Time" and "Steady Time".
+
+We use the following diagrams to visualize the derivation for these conversion equations.
+Steady Time is marked as timeline `S`, and ROS Time as timeline `R`.
+Any `R_n` and `S_n` are matched time points, they coincide with each other: at Steady Time `S_n`, `now()` will return `R_n`.
+Likewise, when `sleep_until(R_n)` is called, the `PlayerClock` which uses steady time for sleeping, will sleep until `S_n`.
+
+In this diagram you can see `rate` changes, `pause`, and `jump`, with arrows showing matched points.
+The bag we are playing starts at R0 and goes on forward in time (we do not show the end of the bag).
+The user is performing actions in `S`, the timeline that they experience.
+
+The lines and numbered references indicate "events" where a time control operation has been performed, such as changing the rate, pausing, or jumping.
+
+Events in the first timeline "Rate Change":
+1. At `S_0`: start playback at rate `0.5`
+1. At `S_1`: change rate to `2.0`
+1. At `S_2`: change rate to `1.0`
+
+Events in the second timeline "Pause / Jump":
+1. At `S_0`: start playback at regular speed
+1. At `S_1`: call `pause`
+1. At `S_2`: call `resume`
+1. At `S_3`: call `jump` to `R_4`
+  * We could also jump backwards on the ROS Timeline just as simply (however that diagram has crossing lines)
+
+![diagram showing two timelines with rate change pause and jump](time_control_timelines.png "Timelines")
+
+In between the time control events, there are arbitrary points `R_x / S_x`, `R_y / S_y`, and `R_z / S_z` that we need to convert between.
+We do that using the following equations:
+
+![equations for converting letter-subscripted time points](time_control_intermediate_equations.png "Intermediate Equations")
+
+The above equations are clearly of a general form, as follows
+
+![final equations for converting time points](time_control_final_equations.png "General Equation")
+
+
+Based on this, the implementation will need to take a "snapshot" of `R_ref` and `S_ref` whenever a time control operation is performed, so that future conversions can be accurately calculated.
```

Comment:
It is a sort of integration, yes, only in the ROS Time point - Steady Time is always measured against a constant reference. Given the fact that we can perform many time control operations during playback, I have not yet seen a possible way to do the time conversion in either direction without this integrated ROS Time point.

If you were to propose alternative equations that can handle the necessary cases, that would be fine.

As-is, we cannot guarantee deterministic playback on a non-realtime operating system anyways. The amount of error accumulation will be extremely small, it should be less than a nanosecond per time control operation, depending where the floating point math rounds to - I would be interested in how you think we could measure this error accumulation, I'm not sure exactly what reference points it would use.

Reply:
Also a note - doesn't your implementation also rely on the same type of integration? I'm looking at this line - it's not using exact references, it's using times between messages and accumulating potential error, but on every single message instead of only on time control events - no?

https://github.com/ros2/rosbag2/pull/680/files#diff-547fe4e718fdbdb2aa48c572ef5fa0706144c8e90e47890c50199f7d3834ae80R232

Reply:
@emersonknapp Based on your answer
> We only advance moving reference on a time control event
The current design should be good enough to go and accumulated errors shouldn't be significant. \

To be clear I am not worry about accumulation of the floating point arithmetic rounding errors. 
I was more worry about the fact the there are some overhead exists when we are trying to `sleep_until` or take current timestamp and translate it. This overhead mostly caused by the implementation specific i.e. locking mutexes, context switching, waiting more than we expecting and doing some extra work after taking current time. \
This overhead accumulation going to be insignificant if we will not advance moving reference for each played back message.

However I am going to come back to my implementation to analyse if we could reuse some approach. I just didn't have time for it today.